### PR TITLE
PLT-2782 | Janus0.6

### PR DESCRIPTION
**Pull request by aarshi0301**
## Change description

> Description here
- optimise cassandra calls
- prefetch all the properties of vertex 
- prefetch all the properties of relation

Atlan PR : https://github.com/atlanhq/atlan/pull/6024/files

stats : https://docs.google.com/spreadsheets/d/1uRahRlM32EKlqOR5c_a4ZB72z9_deXN_CqBS1fbkL2Y/edit?gid=0#gid=0

## Testing Notes

1. Validated contracts
Screenshot 2024-12-09 at 6 27 54 PM

3. Validated McMonitor
Screenshot 2024-12-09 at 5 34 21 PM

4. Validated Business Attributes
Screenshot 2024-12-09 at 5 36 13 PM
 
5. Validated Columns

Screenshot 2024-12-09 at 5 41 24 PM

6. Local

Table
```
"entities": [
        {
            "typeName": "Table",
            "attributes": {
                "lastSyncRunAt": 0,
                "connectionQualifiedName": "default/snowflake/1",
                "__modifiedBy": "admin",
                "columns": [
                    {
                        "guid": "3209048c-bfaf-4304-b6fe-05ec47801f0c",
                        "typeName": "Column",
                        "attributes": {
                            "popularityScore": 1.17549435E-38,
                            "connectorName": "snowflake"
                        },
                        "uniqueAttributes": {
                            "qualifiedName": "table304"
                        }
                    }
                ],
                "__hasLineage": false,
                "adminUsers": [
                    "aarshi",
                    "suman"
                ],
                "queryPreviewConfig": {
                    "env": "[test1, test2]"
                },
                "ownerGroups": [],
                "sourceUpdatedAt": 0,
                "BNEa1vMTtmgZqkTh2icsJB.B8vVPtZI9lbGokUtFjcft4": [
                    "cql",
                    "sql",
                    "test1",
                    "tset2"
                ],
                "announcementUpdatedAt": 0,
                "__timestamp": 1733825739630,
                "sourceCreatedAt": 0,
                "adminRoles": [],
                "adminGroups": [],
                "qualifiedName": "table304",
                "__state": "ACTIVE",
                "sourceReadRecentUserRecordList": [
                    {
                        "typeName": "PopularityInsights",
                        "attributes": {
                            "recordTotalUserCount": 0,
                            "recordUser": "ATLANADMIN",
                            "recordQueryCount": 0,
                            "recordWarehouse": null,
                            "recordComputeCost": 0.0,
                            "recordLastTimestamp": 1693979477505,
                            "recordQuery": null,
                            "recordQueryDuration": 0,
                            "recordMaxComputeCost": 0.0,
                            "recordComputeCostUnit": null
                        }
                    }
                ],
                "dataContractLatest": {
                    "guid": "cd896762-eb74-4742-9825-7c855811506f",
                    "typeName": "DataContract",
                    "attributes": {
                        "popularityScore": 1.17549435E-38
                    },
                    "uniqueAttributes": {
                        "qualifiedName": "table304/Table/contract/V1"
                    }
                },
                "meanings": [],
                "__guid": "ec97889d-f07d-4ec7-98c6-a8badbb79b0d",
                "__createdBy": "admin",
                "__modificationTimestamp": 1733825834082,
                "name": "table304",
                "certificateUpdatedAt": 0
            },
            "guid": "ec97889d-f07d-4ec7-98c6-a8badbb79b0d",
            "status": "ACTIVE",
            "displayText": "table304",
            "classificationNames": [],
            "classifications": [],
            "meaningNames": [],
            "meanings": [],
            "isIncomplete": false,
            "labels": [],
            "createdBy": "admin",
            "updatedBy": "admin",
            "createTime": 1733825739630,
            "updateTime": 1733825834082
        }
    ],
```

MCMonitor

```
"entities": [
        {
            "typeName": "MCMonitor",
            "attributes": {
                "sourceCreatedAt": 0,
                "lastSyncRunAt": 0,
                "adminRoles": [],
                "__modifiedBy": "admin",
                "adminGroups": [],
                "qualifiedName": "monitor1",
                "__state": "ACTIVE",
                "__hasLineage": false,
                "sourceReadRecentUserRecordList": [],
                "meanings": [],
                "__guid": "01b22fa6-0fb3-490f-a6aa-b51d14906297",
                "adminUsers": [],
                "mcMonitorRuleScheduleConfig": {
                    "typeName": "MCRuleSchedule",
                    "attributes": {
                        "mcRuleScheduleType": null,
                        "mcRuleScheduleStartTime": 0,
                        "mcRuleScheduleCrontab": null,
                        "mcRuleScheduleIntervalInMinutes": 0
                    }
                },
                "ownerGroups": [],
                "sourceUpdatedAt": 0,
                "BNEa1vMTtmgZqkTh2icsJB.B8vVPtZI9lbGokUtFjcft4": [],
                "__createdBy": "admin",
                "announcementUpdatedAt": 0,
                "__modificationTimestamp": 1733739725460,
                "name": "monitor1",
                "certificateUpdatedAt": 0,
                "__timestamp": 1733739725460
            },
            "guid": "01b22fa6-0fb3-490f-a6aa-b51d14906297",
            "status": "ACTIVE",
            "displayText": "monitor1",
            "classificationNames": [],
            "classifications": [],
            "meaningNames": [],
            "meanings": [],
            "isIncomplete": false,
            "labels": [],
            "createdBy": "admin",
            "updatedBy": "admin",
            "createTime": 1733739725460,
            "updateTime": 1733739725460
        }
    ]
```

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java [hunk 1]
```diff
@@ -1709,6 +1893,44 @@ public Object getVertexAttribute(AtlasVertex vertex, AtlasAttribute attribute) t
...
+            return properties.get(attribute.getName());
+        }
+
+        //when value is not present and type is primitive, return null
+        if(properties.get(attribute.getName()) == null && isPrefetchValueFinal) {
+            return null;
+        }
+
+        // if value is empty && element is array of primitives, return empty list
+        if (properties.get(attribute.getName()) == null && isArrayOfPrimitives) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java:1918]: This 'if" check will never be true.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java [hunk 2]
```diff
@@ -1093,6 +1178,105 @@ private AtlasEntityHeader mapVertexToAtlasEntityHeader(AtlasVertex entityVertex,
...
+
+            String state = (String)properties.get(Constants.STATE_PROPERTY_KEY);
+            Id.EntityState entityState = state == null ? null : Id.EntityState.valueOf(state);
+            ret.setStatus((entityState == Id.EntityState.DELETED) ? AtlasEntity.Status.DELETED : AtlasEntity.Status.ACTIVE);
+
+            RequestContext context = RequestContext.get();
+            boolean includeClassifications = context.includeClassifications();
+            boolean includeClassificationNames = context.isIncludeClassificationNames();
+            if(includeClassifications){
+                ret.setClassificationNames(getAllTraitNamesFromAttribute(entityVertex));
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java:1206]: We are setting the same values in 'if" and "else"

Comment (mehtaanshul) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java:1206]: Please verify.

Comment (aarshi0301) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java:1206]: Thats snippet from old code, just did copy paste

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java [hunk 3]
```diff
@@ -1063,6 +1053,16 @@ private Map<String, Object> preloadProperties(AtlasVertex entityVertex, AtlasEnt
...
     }
 
+    private void updateAttrValue( Map<String, Object> propertiesMap, VertexProperty<Object> property){
+        Object value = propertiesMap.get(property.key());
+        if (value instanceof List) {
+            ((List) value).add(property.value());
+        } else {
+            List<Object> values = new ArrayList<>();
+            values.add(property.value());
+            propertiesMap.put(property.key(), values);
```
Comment (sriram-atlan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java:1063]: above is not the same as 
` List<Object> values = new ArrayList<>();
                            values.add(propertiesMap.get(property.key()));
                            values.add(property.value());
                            propertiesMap.put(property.key(), values);`

**Review by mehtaanshul — APPROVED** [review]

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java [hunk 4]
```diff
@@ -1709,6 +1897,45 @@ public Object getVertexAttribute(AtlasVertex vertex, AtlasAttribute attribute) t
...
+        //when value is not present and type is primitive, return null
+        if(properties.get(attribute.getName()) == null && isPrefetchValueFinal) {
+            return null;
+        }
+
+        // value is present as marker, fetch the value from the vertex
+        if (ATLAS_INDEXSEARCH_ENABLE_FETCHING_NON_PRIMITIVE_ATTRIBUTES.getBoolean()) {
+            //AtlasPerfMetrics.MetricRecorder nonPrimitiveAttributes = RequestContext.get().startMetricRecord("processNonPrimitiveAttributes");
+            return mapVertexToAttribute(vertex, attribute, null, false);
+            //LOG.debug("capturing excluded property set category and value, mapVertexValue - {}: {} : {} : {}", attribute.getName(), attribute.getAttributeType().getTypeCategory(), properties.get(attribute.getName()), mappedVertex);
```
Comment (sriram-atlan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java:1931]: remove commented lines before merging to master

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java [hunk 5]
```diff
@@ -1009,7 +1019,85 @@ private AtlasEntityHeader mapVertexToAtlasEntityHeader(AtlasVertex entityVertex)
...
+                AccessControlUtils.ATTR_POLICY_ROLES,
+                AccessControlUtils.ATTR_POLICY_ACTIONS,
+                AccessControlUtils.ATTR_POLICY_CATEGORY,
+                AccessControlUtils.ATTR_POLICY_SUB_CATEGORY,
+                AccessControlUtils.ATTR_POLICY_RESOURCES,
+                AccessControlUtils.ATTR_POLICY_IS_ENABLED,
+                AccessControlUtils.ATTR_POLICY_RESOURCES_CATEGORY,
+                AccessControlUtils.ATTR_POLICY_SERVICE_NAME,
+                AccessControlUtils.ATTR_POLICY_PRIORITY,
+                AccessControlUtils.REL_ATTR_POLICIES));
```
Comment (sriram-atlan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java:1084]: I think 3 policy attrs are missing from other branch

**Review by sriram-atlan — DISMISSED** [review]

**Review by sriram-atlan — APPROVED** [review]

**Pull request merged**